ENT-9875: New network parameters

tomstark99 opened the pull request:
- Added `recoveryMaximumBackupInterval: Duration? = null`
- Added `confidentialIdentityMinimumBackupInterval: Duration? = null`

These new parameters are nullable and by default are set to `null` if they are left out of the constructor. This means nothing changes/breaks in the node-api

I have added an additional `DeprecatedConstructorForDeserialization` since the constructor def has been changed and the API diff check did not recognise the ability to use the old constructor with the new param defaults.

I have added another `copy` method to allow calling `NetworkParameters.copy(...)` without specifying the new arguments in the constructor.

These parameters are of type `Duration`, the same as `eventHorizion` which is a consideration that needs to be made when creating a `NetworkParameters` class with the new params.

Commit:
ENT-9875: Added new network parameters

- Added `transactionRecoveryPeriod`
- Added `confidentialIdentityPreGenerationPeriod`

These new parameters are currently set to be nullable meaning they can be ignored and the duration if not specified will be null rather than, e.g., 0. This currently allows for nothing changing/breaking in the node-api

_Note: if these params can stay as nullable then the deprecated constructor might not even be needed (since the existing one will still work), needs to be discussed._

Commit:
detekt fixes

Commit:
Removed redundant constructor

Commit:
Added back LongParameterList

Commit:
Added Deprecated Constructor

- Added a DeprecatedConstructorForSerialization(version=3) to satisfy the api diff scan

colljos reviewed (changes requested):
Recommend adding some tests to associated `NetworkParametersTest`.

colljos commented on core/src/main/kotlin/net/corda/core/node/NetworkParameters.kt:
```diff
@@ -34,10 +34,16 @@ import java.util.Collections.unmodifiableMap
  * @property eventHorizon Time after which nodes will be removed from the network map if they have not been seen
- * during this period
+ * during this period.
+ * @property transactionRecoveryPeriod Time period in days for how far back transactions will be recovered. The assumption is that
+ * transaction prior to this time have already been backed up.
+ * during this period.
+ * @property confidentialIdentityPreGenerationPeriod Time period in days for when Confidential Identities keys were last backed up. The
```
"period in days" => drop "days"?
As the type is a Duration you can pass in other Duration specifications (e.g period in hours might be useful).

Commit:
Updated NetworkMapTest

- Added new parameters to the test NetworkParameter object to make sure network map tests work with new params
- Added a change in new params during one of the flag day tests
- Removed 'days' wording under class documentation

Commit:
Updated parameter names

- changed parameter names to make more sense
- added a test to NetworkParametersTest

colljos reviewed (dismissed):
LGTM!

rick-r3 commented on core/src/main/kotlin/net/corda/core/node/NetworkParameters.kt:
```diff
@@ -34,10 +34,15 @@ import java.util.Collections.unmodifiableMap
  * @property eventHorizon Time after which nodes will be removed from the network map if they have not been seen
- * during this period
+ * during this period.
+ * @property recoveryMaximumBackupInterval Time period for how far back transactions will be recovered. The assumption is that
```
`recoveryMaximumBackupInterval` is a default that will be used by the recovery flows to set how far back in time to consider for recovery.  The expectation is that a node will restore to a database backup no older than this, by default, when attempting recovery.  It can be overridden by specifying an override to the flow.  It will also be overridden if the setting per-node in the node configuration is specified.  An override to the flow takes priority in terms of overrides.

rick-r3 commented on core/src/main/kotlin/net/corda/core/node/NetworkParameters.kt:
```diff
@@ -34,10 +34,15 @@ import java.util.Collections.unmodifiableMap
+ * during this period.
+ * @property recoveryMaximumBackupInterval Time period for how far back transactions will be recovered. The assumption is that
+ * transaction prior to this time have already been backed up.
+ * @property confidentialIdentityMaximumBackupInterval Time period for when Confidential Identities keys were last backed up. The
```
`confidentialIdentityMaximumBackupInterval` is the minimum age of a generated confidential identity before it can be used, by default, unless overridden in the node configuration or a more recent database backup is indicated via RPC / shell.   It is optional in both network parameters and node configuration and if no value is set for either, is assumed to be zero (open to negotiation.... )

However, that description suggests the name is wrong.

`confidentialIdentityMinimumBackupInterval` would be better.

Commit:
Updated parameter name

Commit:
Updated kdoc for parameters

- added more information to try and succinctly describe them and how they are used
